File: packages/core/src/headers.ts
import type { AgentCapabilityScore } from './types.js'

export const AGENTAUTH_HEADERS = {
  STATUS: 'AgentAuth-Status',
  SCORE: 'AgentAuth-Score',
  MODEL_FAMILY: 'AgentAuth-Model-Family',
  POMI_CONFIDENCE: 'AgentAuth-PoMI-Confidence',
  CAPABILITIES: 'AgentAuth-Capabilities',
  VERSION: 'AgentAuth-Version',
  CHALLENGE_ID: 'AgentAuth-Challenge-Id',
  TOKEN_EXPIRES: 'AgentAuth-Token-Expires',
} as const

export type AgentAuthHeaderName = (typeof AGENTAUTH_HEADERS)[keyof typeof AGENTAUTH_HEADERS]

export function formatCapabilities(score: AgentCapabilityScore): string {
  return `reasoning=${score.reasoning},execution=${score.execution},autonomy=${score.autonomy},speed=${score.speed},consistency=${score.consistency}`
}

export function parseCapabilities(header: string): Partial<AgentCapabilityScore> {
  const result: Record<string, number> = {}
  for (const part of header.split(',')) {
    const [key, val] = part.split('=')
    if (key && val) {
      result[key.trim()] = parseFloat(val.trim())
    }
  }
  return result as Partial<AgentCapabilityScore>
}

export function buildHeaders(params: {
  status: 'verified' | 'unverified' | 'expired'
  score?: AgentCapabilityScore
  model_family?: string
  pomi_confidence?: number
  challenge_id?: string
  expires_at?: number
}): Record<string, string> {
  const headers: Record<string, string> = {
    [AGENTAUTH_HEADERS.STATUS]: params.status,
    [AGENTAUTH_HEADERS.VERSION]: '1',
  }

  if (params.score) {
    const avg = (params.score.reasoning + params.score.execution + params.score.autonomy + params.score.speed + params.score.consistency) / 5
    headers[AGENTAUTH_HEADERS.SCORE] = avg.toFixed(2)
    headers[AGENTAUTH_HEADERS.CAPABILITIES] = formatCapabilities(params.score)
  }

  if (params.model_family) {
    headers[AGENTAUTH_HEADERS.MODEL_FAMILY] = params.model_family
  }

  if (params.pomi_confidence !== undefined) {
    headers[AGENTAUTH_HEADERS.POMI_CONFIDENCE] = params.pomi_confidence.toFixed(2)
  }

  if (params.challenge_id) {
    headers[AGENTAUTH_HEADERS.CHALLENGE_ID] = params.challenge_id
  }

  if (params.expires_at) {
    headers[AGENTAUTH_HEADERS.TOKEN_EXPIRES] = params.expires_at.toString()
  }

  return headers
}

File: sdks/python/src/xagentauth/headers.py
from __future__ import annotations

from xagentauth.types import AgentCapabilityScore

AGENTAUTH_HEADERS = {
    "STATUS": "AgentAuth-Status",
    "SCORE": "AgentAuth-Score",
    "MODEL_FAMILY": "AgentAuth-Model-Family",
    "POMI_CONFIDENCE": "AgentAuth-PoMI-Confidence",
    "CAPABILITIES": "AgentAuth-Capabilities",
    "VERSION": "AgentAuth-Version",
    "CHALLENGE_ID": "AgentAuth-Challenge-Id",
    "TOKEN_EXPIRES": "AgentAuth-Token-Expires",
}


def format_capabilities(score: AgentCapabilityScore) -> str:
    """Format capability scores as a comma-separated key=value string.

    Example: "reasoning=0.9,execution=0.85,autonomy=0.8,speed=0.75,consistency=0.88"
    """
    return (
        f"reasoning={score.reasoning},"
        f"execution={score.execution},"
        f"autonomy={score.autonomy},"
        f"speed={score.speed},"
        f"consistency={score.consistency}"
    )


def parse_capabilities(header: str) -> dict[str, float]:
    """Parse a capabilities header string into a dict of dimension -> score.

    Example: "reasoning=0.9,execution=0.85" -> {"reasoning": 0.9, "execution": 0.85}
    """
    result: dict[str, float] = {}
    for part in header.split(","):
        if "=" in part:
            key, val = part.split("=", 1)
            try:
                result[key.strip()] = float(val.strip())
            except ValueError:
                continue
    return result
